robot/impl/local_robot_test.go

```diff
@@ -3238,3 +3240,146 @@ func TestImplicitDepsAcrossModules(t *testing.T) {
 	_, err = r.ResourceByName(motor.Named("m2"))
 	test.That(t, err, test.ShouldBeNil)
 }
+
+func TestResourceByNameAcrossRemotes(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+
+	// Setup a robot1 -> robot2 -> robot3 -> robot4 remote chain. Ensure that if
+	// robot4 has an encoder "e", all robots in the chain can retrieve it by
+	// simple name "e" or short name "[remote-prefix]:e". Also ensure that a
+	// motor "m1" on robot1 can depend on "robot2:robot3:robot4:e" and a motor
+	// "m2" on robot2 can depend on "e".
+
+	startWeb := func(r robot.LocalRobot) string {
+		var boundAddress string
+		for i := 0; i < 10; i++ {
+			port, err := utils.TryReserveRandomPort()
+			test.That(t, err, test.ShouldBeNil)
+
+			options := weboptions.New()
+			boundAddress = fmt.Sprintf("localhost:%v", port)
+			options.Network.BindAddress = boundAddress
+			if err := r.StartWeb(ctx, options); err != nil {
+				r.StopWeb()
+				if strings.Contains(err.Error(), "address already in use") {
+					logger.Infow("port in use; restarting on new port", "port", port, "err", err)
+					continue
+				}
+				t.Fatalf("StartWeb error: %v", err)
+			}
+			break
+		}
+		return boundAddress
+	}
```

Comment:
feels like this loop should be a utility function since we use it more than one place. can be a follow up though

Reply:
Yeah it's an interesting one; I buy that this should be a test utility like `TryStartWebOnRandomPort` or something. The thing is we already have a test utility [`WaitForServing`](https://github.com/viamrobotics/rdk/blob/main/testutils/robottestutils/robot_utils.go#L98-L125) that looks through observed logs for `serving {url: "localhost:[port]"}`, but it interestingly didn't quite work for this use case (something about only looking at `log.Message` instead of looking in the log's _fields_ for the `url: "localhost:[port]"` pair). FWIW @dgottlieb had a couple different strategies for starting a robot on a random port. I copied this pattern from [one](https://github.com/viamrobotics/rdk/commit/6049c6d73f0906623793a1ff61a56be7bc8a8c96#diff-4eff1b8e0794fc6f4b214d99f22637304400bc651db97c8280fdd67914ab9db2R52-R63), but [the other strategy](https://github.com/viamrobotics/rdk/commit/6049c6d73f0906623793a1ff61a56be7bc8a8c96#diff-4eff1b8e0794fc6f4b214d99f22637304400bc651db97c8280fdd67914ab9db2R100-R109) does use `WaitForServing`. I may have to think a bit more, but it'd be nice if there were a single test utility for starting a robot on a random port, and we didn't have to write any `for` loops or `utils.TryReserveRandomPort` calls in the tests (only in the utility).

Reply:
Will continue to think about this and may file a follow up ticket.